model/src/order.rs

```diff
@@ -258,21 +258,70 @@ impl OrderCreation {
     }
 }
 
-// /// An order cancellation as provided to the orderbook by the frontend.
-// #[serde_as]
-// #[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize, Hash)]
-// #[serde(rename_all = "camelCase")]
-// pub struct OrderCancellation {
-//     pub uid: OrderUid,
-//     pub signature: Signature,
-// }
+/// An order cancellation as provided to the orderbook by the frontend.
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize, Hash)]
+pub struct OrderCancellation {
+    // TODO - does it make sense to include the orderUid here?
+    pub order_uid: OrderUid,
+    pub signature: Signature,
+}
+
+impl OrderCancellation {
+    // keccak256("OrderCancellation(bytes orderUid)")
+    pub const ORDER_CANCELLATION_TYPE_HASH: [u8; 32] =
+        hex!("7b41b3a6e2b3cae020a3b2f9cdc997e0d420643957e7fea81747e984e47c88ec");
+
+    fn order_cancellation_digest(&self) -> [u8; 32] {
+        let mut hash_data = [0u8; 64];
+        hash_data[0..32].copy_from_slice(&Self::ORDER_CANCELLATION_TYPE_HASH);
+        hash_data[44..64].copy_from_slice(&signing::keccak256(&self.order_uid.0));
+        signing::keccak256(&hash_data)
+    }
+    // TODO - Many duplicated methods from OrderCreation, need EIP712 Signing Library
```

Comment:
I think these methods could become part of the Signature type and be shared between OrderCreation and OrderCancellation if it takes as argument the domain separator and `hash_data`. `hash_data` can then be composed individually inside `OrderCancellation` and `OrderCreation`

Reply:
Yes certainly. I guess this would be an intermediary step between a fully compatible EIP712 signing library.

Reply:
Have made an EIP712Signing trait that is used by both `OrderCreation` and `OrderCancellation` which now only have to implement their unique "digest".

Reply:
I would prefer doing this in this PR (or even better throwing in a PR before that moves the existing order signature validation into the Signature struct) over introducing the redundancy here.